rfd/0186-optionally-require-reason-for-access-request.md

```diff
@@ -0,0 +1,215 @@
+---
+authors: Pawel Kopiczko (pawel.kopiczko@goteleport.com)
+state: draft
+---
+
+# RFD 186 - Optionally require reason for access request
+
+
+## Required Approvers
+
+* Engineering: @r0mant && @smallinsky
+* Product: @klizhentas && @xinding33
+
+
+## What
+
+Add the ability to set a policy to make a reason required for an access request
+to be submitted
+([#20164](https://github.com/gravitational/teleport/issues/20164)).
+
+
+## Why
+
+Improves user experience for environments with audit/compliance policies that
+require a reason. Adding the ability to mark a reason as required would allow
+both tsh and the web UI to prompt for a reason without the end user needing to
+remember that it's required and as a result reduce the number of invalid access
+requests that need to be rejected.
+
+
+## Details
+
+
+### Possible implementations
+
+
+#### 1. Add a new value to role.options.request_access
+
+Currently a role can have `request_access` option set to one of the following
+values: `optional`, `always` and `reason` as described in
+https://goteleport.com/docs/admin-guides/access-controls/access-requests/access-request-configuration/#how-clients-request-access.
+
+The idea is to support another value `reason-required`.
+
+> [!NOTE]
+> role.options.request_access is a global configuration
+
+It has to be noted that role.options.request_access has a global scope. I.e. if
+it's set on any role all access requests to all roles and resources are
+affected.
+
+Example:
+
+For two roles `rr-role` and `dba-req-role` as defined below:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: rr-role
+spec:
+  options:
+    request_access: reason-required
+```
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: dba-req-role
+spec:
+  allow:
+    request:
+      roles:
+        - 'dba'
+```
+
+A user with both roles `rr-role` and `dba-req-role` assigned will be affected
+by request_access set in role `rr-role` while requesting access to role `dba`
+allowed by role `dba-req-role`.
+
+Because request_access option is global a priority has to be establish when
+more than one role has the option specified:
+
+- reason
+- always
+- reason-required
+- optional (default)
+
+Problems:
+
+- It may be confusing/undesired that roles with request_access option set to
+  "reason-required" affects other roles.
```

Comment:
The main problem I have this is that `request_access: reason` means an access request is automatically "required" on every single login, and `request_access: reason-required` means something completely different, now a reason is required on optional access requests made later. And that's very confusing, opaque, and almost seems like it would be better if the names were swapped